kore/src/Kore/Domain/Builtin.hs

```diff
@@ -205,110 +143,350 @@ instance Unparse child => Unparse (InternalList child) where
+instance (Debug key, Debug (valueWrapper child), Debug child)
+    => Debug (NormalizedAc key valueWrapper child)
 
-emptyNormalizedSet :: NormalizedSet key child
-emptyNormalizedSet = NormalizedSet
+emptyNormalizedAc :: NormalizedAc key valueWrapper child
+emptyNormalizedAc = NormalizedAc
     { elementsWithVariables = []
-    , concreteElements = Set.empty
-    , sets = []
+    , concreteElements = Map.empty
+    , opaque = []
     }
 
-{- | Internal representation of the builtin @SET.Set@ domain.
- -}
-data InternalSet key child =
-    InternalSet
-        { builtinSetSort :: !Sort
-        , builtinSetUnit :: !Symbol
-        , builtinSetElement :: !Symbol
-        , builtinSetConcat :: !Symbol
-        , builtinSetChild :: !(NormalizedSet key child)
+{- | Internal representation of associative-commutative domain values.
+-}
+data InternalAc key (normalized :: * -> * -> *) child =
+    InternalAc
+        { builtinAcSort :: !Sort
+        , builtinAcUnit :: !Symbol
+        , builtinAcElement :: !Symbol
+        , builtinAcConcat :: !Symbol
+        , builtinAcChild :: normalized key child
         }
     deriving (Eq, Foldable, Functor, Traversable, GHC.Generic, Ord, Show)
 
-instance (Hashable key, Hashable child) => Hashable (InternalSet key child)
+{- | Establishes a bijection between value wrappers and entire-structure
+wrappers, with a few utility functions for the two.
+-}
+class AcWrapper (normalized :: * -> * -> *) (valueWrapper :: * -> *)
+    | normalized -> valueWrapper
+    , valueWrapper -> normalized
+  where
+    unwrapAc :: normalized key child -> NormalizedAc key valueWrapper child
+    wrapAc :: NormalizedAc key valueWrapper child -> normalized key child
+
+    {-| Pairs the values in two wrappers as they should be paired for
+    unification.
+    -}
+    acExactZip
+        :: valueWrapper a -> valueWrapper b -> Maybe (valueWrapper (a, b))
+    unparseElement
+        :: (key -> Pretty.Doc ann)
+        -> (child -> Pretty.Doc ann)
+        -> (child, valueWrapper child) -> Pretty.Doc ann
+    unparseConcreteElement
+        :: (key -> Pretty.Doc ann)
+        -> (child -> Pretty.Doc ann)
+        -> (key, valueWrapper child) -> Pretty.Doc ann
+
+unparsedChildren
+    :: forall ann child key normalized valueWrapper
+    .  (AcWrapper normalized valueWrapper)
+    => (key -> Pretty.Doc ann)
+    -> (child -> Pretty.Doc ann)
+    -> normalized key child
+    -> [Pretty.Doc ann]
+unparsedChildren keyUnparser childUnparser wrapped =
+    case unwrapped of
+        -- case statement needed only for getting compiler notifications when
+        -- the NormalizedAc field count changes
+        NormalizedAc _ _ _ ->
+            (elementUnparser <$> elementsWithVariables)
+            ++ (concreteElementUnparser <$> Map.toAscList concreteElements)
+            ++ (argument' . childUnparser <$> opaque)
+  where
+    unwrapped :: NormalizedAc key valueWrapper child
+    unwrapped = unwrapAc wrapped
```

Comment:
We could also write,
```.hs
    unwrapped@(NormalizedAC _ _ _) = ...
```
to remove the `case` and clean up the function body a bit.